### PR TITLE
chore: add error message

### LINKED ISSUE #322: Check for docutils 0.18 and 0.19 and raise a helpful warning if it is installed
As we've discovered in https://github.com/mcmtroffaes/sphinxcontrib-bibtex/issues/309, there is a nasty docutils bug in 0.18 and 0.19 that will break this extension in hidden ways (where people won't get an error, the page will just be wrong).

Since 0.20 isn't released yet, could we add a short-term quality of life fix to raise an error and give users guidance? I imagine a little function that triggers on `config-inited` and does the following:

1. Check if the user has specified configuration for this extension (meaning they intend to use it)
2. If so, check if docutils is `0.18 >= docutils version < 0.20`
3. If so, then raise a _Warning_ saying "This extension is not compatible with docutils versions 0.18 and 0.20, please pin `docutils==0.17` and `sphinx==5` until docutils 0.20 is released."

It could be a lightweight way to point people in the right direction if they hit this bug, given that getting the right versions installed will require a bit of searching.

### PR DESCRIPTION
**Pull request by agoose77**
Fixes #322, with an open ended version bound (users will be required to upgrade this package in addition to docutils).

### CODE REVIEW

File: src/sphinxcontrib/bibtex/__init__.py [hunk 1]
```diff
@@ -48,7 +53,21 @@ def setup(app: Sphinx) -> Dict[str, Any]:
     app.add_post_transform(BibliographyTransform)
     app.add_domain(BibtexFootDomain)
     app.add_directive("footbibliography", FootBibliographyDirective)
-    app.add_role("footcite", FootCiteRole())
+    app.add_role("footcite", FootCiteRole())"config",
+        )
+
+    # Catch bug in newer docutils
+    if docutils.__version_info__.major == 0 and docutils.__version_info__.minor >= 18:
```
Comment (mcmtroffaes) [inline src/sphinxcontrib/bibtex/__init__.py:60]: Two comments here:
1. What's the reason for not restricting this to minor < 20? We know the fix is applied to the current git version of docutils, so the next version will very likely include the fix. In that way only docutils needs to be updated.
2. Tuple comparison might simplify the logic slightly.

File: src/sphinxcontrib/bibtex/__init__.py [hunk 2]
```diff
@@ -48,7 +53,21 @@ def setup(app: Sphinx) -> Dict[str, Any]:
     app.add_post_transform(BibliographyTransform)
     app.add_domain(BibtexFootDomain)
     app.add_directive("footbibliography", FootBibliographyDirective)
-    app.add_role("footcite", FootCiteRole())
+    app.add_role("footcite", FootCiteRole())"config",
+        )
```
Comment (mcmtroffaes) [inline src/sphinxcontrib/bibtex/__init__.py:57]: Typo?

**Review by mcmtroffaes — CHANGES_REQUESTED** [review]

File: src/sphinxcontrib/bibtex/__init__.py (hunk 2 above)
Comment (agoose77) [inline src/sphinxcontrib/bibtex/__init__.py:57]:
```suggestion
    app.add_role("footcite", FootCiteRole())
```

Comment (agoose77) [reply src/sphinxcontrib/bibtex/__init__.py]: Ugh, my WM is playing up... yes!

File: src/sphinxcontrib/bibtex/__init__.py (hunk 1 above)
Comment (agoose77) [reply src/sphinxcontrib/bibtex/__init__.py:59]: > What's the reason for not restricting this to minor < 20?

That's true, I hadn't fully appreciated that the fix was already in `main`. Let's fix this.

File: src/sphinxcontrib/bibtex/__init__.py [hunk 3]
```diff
@@ -50,6 +55,19 @@ def setup(app: Sphinx) -> Dict[str, Any]:
     app.add_directive("footbibliography", FootBibliographyDirective)
     app.add_role("footcite", FootCiteRole())
 
+    # Catch bug in newer docutils
+    if docutils.__version_info__.major == 0 and docutils.__version_info__.minor >= 18:
```
Comment (agoose77) [inline src/sphinxcontrib/bibtex/__init__.py:59]:
```suggestion
    if (0, 18, 0) <= docutils.__version_info__[:3]:
```

File: src/sphinxcontrib/bibtex/__init__.py [hunk 4]
```diff
@@ -50,6 +55,19 @@ def setup(app: Sphinx) -> Dict[str, Any]:
     app.add_directive("footbibliography", FootBibliographyDirective)
     app.add_role("footcite", FootCiteRole())
 
+    # Catch bug in newer docutils
+    if (0, 18, 0) <= docutils.__version_info__[:3]:
```
Comment (agoose77) [inline src/sphinxcontrib/bibtex/__init__.py:59]:
```suggestion
    if (0, 18, 0, 'final', 0, True) <= docutils.__version_info__ < (0, 20, 0, 'final', 0, True):
```

File: src/sphinxcontrib/bibtex/__init__.py (hunk 1 above)
Comment (agoose77) [reply src/sphinxcontrib/bibtex/__init__.py]: (2) yes, I was a bit twitchy about unpacking a tuple like this, but I doubt it will change.

(1) it's an assumption that might be wrong, rather than the guarantee that we'll need to update this package. Both are fine, we can always re-release this package if there turns out to be an issue.

File: src/sphinxcontrib/bibtex/__init__.py [hunk 5]
```diff
@@ -50,6 +55,19 @@ def setup(app: Sphinx) -> Dict[str, Any]:
     app.add_directive("footbibliography", FootBibliographyDirective)
     app.add_role("footcite", FootCiteRole())
 
+    # Catch bug in newer docutils
+    if (0, 18, 0, 'final', 0, True) <= docutils.__version_info__ < (0, 20, 0, 'final', 0, True):
```
Comment (mcmtroffaes) [inline src/sphinxcontrib/bibtex/__init__.py:59]: ``(0, 18) <= docutils.__version_info__ < (0, 20)`` should do the trick

File: src/sphinxcontrib/bibtex/__init__.py [hunk 6]
```diff
@@ -50,6 +55,19 @@ def setup(app: Sphinx) -> Dict[str, Any]:
...
+        error = RuntimeError(
+            (
+                "sphinxcontrib-bibtex is known to be broken with the installed version of "
+                "docutils ({}). Once a fixed version is available, please update both "
+                "packages (sphinxcontrib-bibtex and docutils) to remove this error."
+            ).format(docutils.__version__)
+        )
+        SPHINX_LOGGER.critical(error.args[0])
+        raise error
```
Comment (mcmtroffaes) [inline src/sphinxcontrib/bibtex/__init__.py:68]: People who already run the extension with docutils 0.18 or 0.19 should be able to keep using the extension as before, in order not to cause a mass breakage. Is it possible just to issue a warning rather than raising an error?

File: src/sphinxcontrib/bibtex/__init__.py [hunk 7]
```diff
@@ -50,6 +55,19 @@ def setup(app: Sphinx) -> Dict[str, Any]:
     app.add_directive("footbibliography", FootBibliographyDirective)
     app.add_role("footcite", FootCiteRole())
 
+    # Catch bug in newer docutils
+    if (0, 18, 0, 'final', 0, True) <= docutils.__version_info__ < (0, 20, 0, 'final', 0, True):
+        error = RuntimeError(
+            (
+                "sphinxcontrib-bibtex is known to be broken with the installed version of "
+                "docutils ({}). Once a fixed version is available, please update both "
+                "packages (sphinxcontrib-bibtex and docutils) to remove this error."
```
Comment (mcmtroffaes) [inline src/sphinxcontrib/bibtex/__init__.py:64]: I'd rephrase this warning to be more clear and specific, e.g.

```Beware that docutils versions 0.18 and 0.19 (you are running {}) are known to generate invalid html for citations. If this issue affects you, please use docutils<=0.17 (or >=0.20 once released) instead. For more details, see https://sourceforge.net/p/docutils/patches/195/```

Also I think this message should only be shown when generating html; no point in bugging users who don't use the html builder.

File: src/sphinxcontrib/bibtex/__init__.py (hunk 5 above)
Comment (agoose77) [reply src/sphinxcontrib/bibtex/__init__.py:59]: Oh, I had forgotten about this feature of tuple comparisons. It will differ for the lower upper bounds (`(0, 18) <= docutils.__version_info__` includes all pre-releases of 0.18, whilst `docutils.__version_info__ < (0, 20)` includes no pre-releases of 0.20) but I think that's what we want here.

Comment (mcmtroffaes) [reply src/sphinxcontrib/bibtex/__init__.py:59]: Yeah, I think so. I believe this is consistent with what ``install_requires`` does (though pip will not include pre-releases unless explicitly instructed), i.e. this comparison should be identical to ``docutils>=0.18,<0.20``.

**Comment by agoose77** [comment]
Hi @mcmtroffaes, I pinged you for a final review, but wanted to write as much in person!

Thanks for your patience on this. I'm sure you understand, I had some other projects take some of my time. I believe that this is ready for merging, does that match your assessment? 

Thanks!

File: src/sphinxcontrib/bibtex/__init__.py [hunk 8]
```diff
@@ -43,16 +47,30 @@ def setup(app: Sphinx) -> Dict[str, Any]:
...
     return {
-        'version': '2.5.1a0',
-        'env_version': 9,
-        'parallel_read_safe': True,
-        'parallel_write_safe': True,
-        }
+        "version": "2.5.1a0",
+        "env_version": 9,
+        "parallel_read_safe": True,
+        "parallel_write_safe": True,
```
Comment (mcmtroffaes) [inline src/sphinxcontrib/bibtex/__init__.py:75]: These changes need to be reverted (it will fix the setup error).

File: src/sphinxcontrib/bibtex/__init__.py [hunk 9]
```diff
@@ -43,16 +47,30 @@ def setup(app: Sphinx) -> Dict[str, Any]:
...
+        < (0, 20, 0, "final", 0, True)
+    ) and app.builder.name == "html":
+        error = RuntimeError(
+            (
+                "Beware that docutils versions 0.18 and 0.19 (you are running {}) are known to generate invalid html "
+                "for citations. If this issue affects you, please use docutils<=0.17 (or >=0.20 once released) instead."
+                " For more details, see https://sourceforge.net/p/docutils/patches/195/"
+            ).format(docutils.__version__)
+        )
+        SPHINX_LOGGER.warn(error.args[0])
```
Comment (mcmtroffaes) [inline src/sphinxcontrib/bibtex/__init__.py:69]: Why not just emit the warning message directly (instead of first wrapping it inside a runtime error)?

Comment (mcmtroffaes) [reply src/sphinxcontrib/bibtex/__init__.py:69]: Also the SPHINX_LOGGER variable appears nowhere to be defined and/or imported.

File: src/sphinxcontrib/bibtex/__init__.py [hunk 10]
```diff
@@ -43,16 +47,30 @@ def setup(app: Sphinx) -> Dict[str, Any]:
...
     app.add_post_transform(BibliographyTransform)
     app.add_domain(BibtexFootDomain)
     app.add_directive("footbibliography", FootBibliographyDirective)
     app.add_role("footcite", FootCiteRole())
 
+    # Catch bug in newer docutils
+    if (
+        (0, 18, 0, "final", 0, True)
+        <= docutils.__version_info__
+        < (0, 20, 0, "final", 0, True)
```
Comment (mcmtroffaes) [inline src/sphinxcontrib/bibtex/__init__.py:60]: Please use tuple comparison against just the numeric part of the version; e.g. ``(0, 18) <= doctutils.__version_info__ < (0, 20)`` should do the trick.

File: src/sphinxcontrib/bibtex/__init__.py [hunk 11]
```diff
@@ -43,16 +47,30 @@ def setup(app: Sphinx) -> Dict[str, Any]:
     app.add_directive("bibliography", BibliographyDirective)
     app.add_role("cite", CiteRole())
     app.add_node(bibliography, override=True)
-    app.add_node(raw_latex, latex=(visit_raw_latex, depart_raw_latex),
-                 override=True)
+    app.add_node(raw_latex, latex=(visit_raw_latex, depart_raw_latex), override=True)
```
Comment (mcmtroffaes) [inline src/sphinxcontrib/bibtex/__init__.py:50]: Please revert this change.

File: src/sphinxcontrib/bibtex/__init__.py [hunk 12]
```diff
@@ -16,6 +17,9 @@
 from .foot_directives import FootBibliographyDirective
 
 
+logger = logging.getLogger(__name__)
+
+
```
Comment (mcmtroffaes) [inline src/sphinxcontrib/bibtex/__init__.py:22]: The logger object is nowhere used; remove?

Comment (mcmtroffaes) [reply src/sphinxcontrib/bibtex/__init__.py:22]: (Or maybe it is meant to be used instead of SPHINX_LOGGER below?)

**Review by mcmtroffaes — CHANGES_REQUESTED** [review]

File: src/sphinxcontrib/bibtex/__init__.py (hunk 8 above)
Comment (agoose77) [reply src/sphinxcontrib/bibtex/__init__.py:75]: Wait, this shouldn't have changed... My local tooling must have been applying black.

File: src/sphinxcontrib/bibtex/__init__.py [hunk 13]
```diff
@@ -50,6 +54,17 @@ def setup(app: Sphinx) -> Dict[str, Any]:
...
+    # Catch bug in newer docutils
+    if ((0, 18) <= docutils.__version_info__ < (0, 20)) and app.builder.name == "html":
+        error = RuntimeError(
+            (
+                "Beware that docutils versions 0.18 and 0.19 (you are running {}) are known to generate invalid html "
+                "for citations. If this issue affects you, please use docutils<=0.17 (or >=0.20 once released) instead."
+                " For more details, see https://sourceforge.net/p/docutils/patches/195/"
+            ).format(docutils.__version__)
+        )
+        logger.warn(error.args[0])
```
[the remark this answers is not in the document]
Comment (agoose77) [reply src/sphinxcontrib/bibtex/__init__.py:66]: Nope, we were raising it at one point I think. Good suggestion.

Sorry, I haven't given this PR as thorough a once-over as I should have done, which is why it's been like pulling teeth! Thanks!

**Comment by mmcky** [comment]
> * The builder is not necessarily already created when setup is being called, or so it seems... see https://github.com/sphinx-doc/sphinx/blob/ba080286b06cb9e0cadec59a6cf1f96aa11aef5a/sphinx/application.py#L256 - this occurs only after the extensions are loaded here: https://github.com/sphinx-doc/sphinx/blob/ba080286b06cb9e0cadec59a6cf1f96aa11aef5a/sphinx/application.py#L222 It looks like this is only an issue for newer versions of sphinx. Given this, I'm not sure yet how to access the buildername from within the setup...

It looks like you could use `buildername` given the `builder` isn't full instantiated at the extension load phase which is passed in  via the `__init__` function (https://github.com/sphinx-doc/sphinx/blob/ba080286b06cb9e0cadec59a6cf1f96aa11aef5a/sphinx/testing/util.py#L103) for both `sphinx` and `SphinxTestApp`

**Comment by mcmtroffaes** [comment]
As I said, I'm not quite sure yet how to use buildername from within the setup. Do you know how @mmcky ?

**Comment by mcmtroffaes** [comment]
... and docutils 0.20 was released; let's use that instead. Closing in favour of #330.
